src/submodules/io/csv_reader.jl

```diff
@@ -12,12 +13,27 @@ to fields of the struct.
 """
 function populate_from_csv!(
     filename,
-    model::ABM{A},
+    model::ABM{S,A},
     agent_type::B;
     kwargs...
-) where {A, B<:Union{Type{<:A}, Function}}
-    !(agent_type isa Function) && !haskey(kwargs, :types) && (kwargs[:types] = collect(fieldtypes(agent_type)))
-    for row in Rows(filename; kwargs...)
+) where {A, B<:Type{<:A},S}
+    !haskey(kwargs, :types) && (kwargs = (kwargs..., types = collect(fieldtypes(agent_type))))
+    for row in CSV.Rows(filename; kwargs...)
         add_agent_pos!(agent_type(row...), model)
     end
 end
+
+function populate_from_csv!(
+    filename,
+    model::ABM{S,A},
+    agent_type::Function;
+    kwargs...
+) where {S, A}
+    for row in CSV.Rows(filename; kwargs...)
+        add_agent_pos!(agent_type(row...), model)
+    end
+end
+
+function Base.parse(::Type{NTuple{N,T}}, s::String) where {N,T}
+    return Tuple(map(x -> parse(T, x), split(s[2:(end-1)], ", ")))
+end
```

Comment:
TYPE PIRACY DETECTED!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Reply:
IM CALLING THE POLICE

Reply:
🚓  👮🏻

Reply:
I didn't know what that was. It seems to be a bad idea. I'll try look for another way around this. Without it, CSV.jl complains about this method not existing

Reply:
Ooooh, then you should open an issue there to support parsing as Tuples? But I'm not sure whether this is intended for CSV. Data there in each column do not typically come as tuples. Why is it the case here? CSV format is quite barebones, don't think you can support "tuples as dcolumns".

Reply:
CSV.jl doesn't allow specifying a different function to parse a specific column. The only solutions I could think of are:
- Create a wrapper for `NTuple`, and define `Base.parse` for that type
- Keeping the Tuple fields as `String`, then once CSV delivers a row convert the required field(s) manually
- Only keeping the `populate_from_csv!` method that takes a `Function`
- Require that `agent_type` have a constructor taking keyword arguments, and have the user manually handle non-primitive types that will be handed over as `String`s

Reply:
You should open a request issue on CSV to allow custom parsing function that defaults to `Base.parse`, that seems straight-frward

Reply:
> Ooooh, then you should open an issue there to support parsing as Tuples? But I'm not sure whether this is intended for CSV. Data there in each column do not typically come as tuples. Why is it the case here? CSV format is quite barebones, don't think you can support "tuples as dcolumns".

For some reason none of your messages after the first seemed to load in until I reloaded the page, sorry for ignoring this.

Agents would require one column to be a tuple for their position. Usually non-scalar fields would span multiple columns, but that would result in multiple fields instead of one tuple.